### ISSUE TITLE
Allow querying and subscribing from a given point in time (`readTime`)

### ISSUE THREAD

**Issue by merlinnot** [googleapis/nodejs-firestore#1457]
**Is your feature request related to a problem? Please describe.**

Firestore SDK allows for cancellation of queries, but does not provide a "resubscribe" mechanism, so that the queries can be "paused". Use of such mechanism would allow to:
- Pause subscriptions to datasets where they are not needed (e.g. a given instance is not serving any requests).
- Cache results of collection queries, then issue a query "changed since" to fetch only the changes since the last cache entry was stored.

I am aware that both of these functionalities can be implemented client-side if developers manually create an "updated at" field in each document, however an index is required on such field and leads to hotspotting. Firestore back ends already expose an API to query for these changes, while not limiting the scalability of a given collection: `readTime`.

 **Describe the solution you'd like**

I'd like to be able to make queries like this:

```ts
firestore
  .collection('collection')
  .changedSince(timestamp)
  .onSnapshot()
```

and this:

```ts
firestore
  .collection('collection')
  .changedSince(timestamp)
  .get()
```

 **Describe alternatives you've considered**

I considered:
- Adding a field to documents and updating it on each write. This leads to hotspotting, as an index on this field is required. It also requires additional development effort (changes to code reading and writing, not only reading).
- Implementing a Firestore client myself, which is what I did, however this API is not trivial and I don't think I'd advise anyone to do that unless they really, really need it (which is the case at my company). I'm proposing to add it to the API, so that the general public can benefit from it.

 **Additional context**

See `read_time` in [the Firestore RPC reference for `Target`](https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#target).

**Comment by wu-hui** [comment]
Thanks for your suggestion! You definitely did good research!

This feature is called query resumption, and it is handled transparently in client SDKs (JS/Android/iOS). They could do it because they save the queries (even after unregister) with the `read_time`. When queries are re-registered, those `read_time`s are also sent to backend. Backend will send only the updates since that `read_time`, *only if* the `read_time` is no longer than 30 minutes ago. Over 30 minutes, actual reads are still being performed, and server will treat the request as a brand new request (without `read_time`). The client SDK has to handle both cases.

The server SDKs cannot do what client SDKs do, because they do not keep a interval cache. Note the backend does not keep a change log forever, so my concern with the api you proposed is that it looks like user can get all writes from any point in time, but this is not feasible. 

Even if we accept `QuerySnapshot` for `changedSince` instead, it might still be out of the 30 minute window, which means it will not work as intended still.

OTOH, have you considered building your own snap cache in your app? This way you do not actually pause a listener, but simply direct the snapshots to the cache?

**Comment by merlinnot** [comment]
> OTOH, have you considered building your own snap cache in your app? This way you do not actually pause a listener, but simply direct the snapshots to the cache?

Yes, this is the solution which I started with, there are however some issues with it:
1) The cache is being updated even if the fresh data is not actually needed (that is, no one is calling my API or at least some instances). Imagine a subscription to a collection in which 10% of documents are updated once per minute, the remaining 90% are updated once per day (this is, of course, a very simplified example). If I have no requests for an hour, I will still be fetching (and therefore paying) for 60 updates to this 10%, where in reality no one will use this data (the same set of documents will be overwritten 60 times).
2. When using Cloud Run, instances are throttled when out of request - although the CPU is not throttled down to 0, there's not enough of CPU time to actually keep the subscription open (been there, done that). This is quite reasonable, that if the instance is not serving any requests, there's also no need to keep the Firestore subscription alive.

> Backend will send only the updates since that read_time, only if the read_time is no longer than 30 minutes ago. Over 30 minutes, actual reads are still being performed, and server will treat the request as a brand new request (without read_time).

I have not observed this behavior, I re-tested it just now. I have a collection in a test project that I modified last week, it has ~100 documents in it. I added one document to it (with ID `the-only-one-new`) and I called the Listen RPC using the following script:

```ts
import { v1 } from '@google-cloud/firestore';

const CLOUD_RESOURCE_HEADER = 'google-cloud-resource-prefix';
const PROJECT_ID = 'my-project';

const client = new v1.FirestoreClient({ projectId: PROJECT_ID });

const database = `projects/${PROJECT_ID}/databases/(default)`;
const parent = `${database}/documents`;
const collectionId = 'my-collection';

const request = {
  addTarget: {
    query: { parent, structuredQuery: { from: [{ collectionId }] } },
    readTime: {
      nanos: 0,
      // This is March 29th, 15:00 UTC - "5 hours ago".
      seconds: 1617030000,
    },
    targetId: 1,
  },
  database,
};

const main = async () => {
  const stream = client.listen({
    otherArgs: { headers: { [CLOUD_RESOURCE_HEADER]: database } },
  });

  stream.write(request);

  stream.on('data', (data) => {
    console.log(data);
  });
};

main();
```

This is the response I got:

```
{
  targetChange: {
    targetIds: [ 1 ],
    targetChangeType: 'ADD',
    cause: null,
    resumeToken: <Buffer >,
    readTime: null
  },
  responseType: 'targetChange'
}
{
  targetChange: {
    targetIds: [],
    targetChangeType: 'NO_CHANGE',
    cause: null,
    resumeToken: <Buffer 0a 09 08 80 f8 f6 95 e2 d5 ef 02>,
    readTime: { seconds: '1617030000', nanos: 0 }
  },
  responseType: 'targetChange'
}
{
  documentChange: {
    targetIds: [ 1 ],
    removedTargetIds: [],
    document: {
      fields: [Object],
      name: 'projects/my-project/databases/(default)/documents/my-collection/the-only-one-new',
      createTime: [Object],
      updateTime: [Object]
    }
  },
  responseType: 'documentChange'
}
{ filter: { targetId: 1, count: 64 }, responseType: 'filter' }
{
  targetChange: {
    targetIds: [ 1 ],
    targetChangeType: 'CURRENT',
    cause: null,
    resumeToken: <Buffer 0a 09 08 e0 f6 c9 82 96 d6 ef 02>,
    readTime: { seconds: '1617043918', nanos: 60384000 }
  },
  responseType: 'targetChange'
}
{
  targetChange: {
    targetIds: [],
    targetChangeType: 'NO_CHANGE',
    cause: null,
    resumeToken: <Buffer 0a 09 08 e0 f6 c9 82 96 d6 ef 02>,
    readTime: { seconds: '1617043918', nanos: 60384000 }
  },
  responseType: 'targetChange'
}
```

As you can see, there's only one `documentChange`. I do understand that the back end might have to read all documents (as the `filter` event contains the count, so I imagine it must _at least_ get the count from each shard and sum it up. However, even if the back end reads these documents internally, it doesn't send them to the client - it sends only documents changed since the provided time (no matter how old `read_time` is - I tested it with a timestamp 1 week old and it still behaved as expected, returning just three out of ~100 documents).

This allows me to "resubscribe" to data in Cloud Run very quickly, since I get only the diff. No matter if it's been just a minute between requests, or the entire night (this particular service where I tested it doesn't have much usage during the night). First user in the morning will have just a minor delay - they fetch only data that changed overnight.

Maybe we could start by verifying the exact behavior of `read_time`? I think Googlers like you are the only people who can do that. All of the API documentation I found didn't provide any details, it was mostly learning through experimenting on my side. I also have plans to open source my "Firestore Extensions" library once I stabilize the API, but if you'd like to take a look at what I'm building (mainly these resumable subscriptions), I'd be happy to give you early access :slightly_smiling_face: Especially that my need originates from Cloud Run, which is a product of Google :wink:

**Comment by merlinnot** [comment]
Tiny correction: I just noticed that we have the count in the `filter`event, so it's not "~100 documents", it's exactly 64 documents. At least you can clearly see that the collection has more documents than just the one returned :slightly_smiling_face:

**Comment by wu-hui** [comment]
You are right. Setting `read_time` will only send you what has updated since then, even out of 30 minute window. The difference I was trying to say is after 30 minutes, actual reads will be performed, therefore even if you only received one document, you still need to pay for the reads in your experiment (64 reads in your case).

However, you could unlisted the query from there on, and resume whenever is needed by your application. I have to stress about the 30 minute window limit there, because I don't want you to be surprised by you bill.

So yeah, what you proposed is definitely possible, but the pricing implication is not clear through the API which you should probably make that clear in your doc if you plan to share it with others.

We do have a feature planned that might address your problem, but we do not know when it could land. We are definitely interested in your work though, and feel free to share it back here. I am closing this issue for now, because there is nothing more I could provide. But feel free to open new issues while you work on your extension. You could also share your extension by opening new issues here, because closed issues might not get enough attention.
